### PR TITLE
support contributes.jsonValidation VSCode API.

### PR DESCRIPTION
**Pull request by bd82**
#### What it does

Support VSCode [contributes.jsonValidation](https://code.visualstudio.com/api/references/contribution-points#contributes.jsonValidation) API.

Note that jsonValidation contribution have a **higher** priority than SchemaStore schemas, and a **lower** priority than registered schemas from the preferences
(In case of identical `fileMatch` strings).

#### How to test

- Create a dummy VSCode ext
- Add in its package.json `contributes.jsonValidation` section pointing to some schema from the schema store with an appropriate `fileMatch` property`.
- Bundle your VSCode ext into a VSIX and place it the `plugins` folder.
- Start Theia
- Inspect that the JSON schema is used from the given

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by bd82** [comment]
Hmm I need to figure out how to sign the commit

File: packages/json/src/browser/json-validation-registry.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+@injectable()
+/**
+ * Holds JSON Validation contributions from plugins / VSCode extensions.
+ * -  https://code.visualstudio.com/api/references/contribution-points#contributes.jsonValidation
+ */
+export class JsonValidationContributionsRegistry {
+
+    protected readonly registry: JsonSchemaConfiguration[] = [];
+
+    registerJsonValidation(jsonValidationEntry: JsonSchemaConfiguration): Disposable {
```
Comment (akosyakov) [inline packages/json/src/browser/json-validation-registry.ts:31]: Why do we need it? There is already `JsonSchemaStore.registerSchema`, please use it.

File: packages/json/src/browser/json-validation-registry.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        this.registry.push(jsonValidationEntry);
+        return Disposable.NULL;
```
Comment (akosyakov) [inline packages/json/src/browser/json-validation-registry.ts:33]: `Disposable.NULL` is bogus, please use `JsonSchemaStore.registerSchema` instead. it has a proper logic to unregster configuration

File: packages/plugin-ext/package.json [hunk 3]
```diff
@@ -11,6 +11,7 @@
     "@theia/editor": "^1.0.0",
     "@theia/file-search": "^1.0.0",
     "@theia/filesystem": "^1.0.0",
+    "@theia/json": "^1.0.0",
```
Comment (akosyakov) [inline packages/plugin-ext/package.json:14]: we don't this dependency, we only need to register configuration in core json schema storage

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 4]
```diff
@@ -139,6 +140,11 @@ export interface PluginPackageSnippetsContribution {
     path?: string;
 }
 
+export interface PluginJsonValidationContribution {
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/plugin-protocol.ts:143]: proper shape: https://github.com/microsoft/vscode/blob/0c25f5a737a0c4b0945783588c13cfee11d0ecb1/src/vs/workbench/api/common/jsonValidationExtensionPoint.ts#L12-L15

File: packages/plugin-ext/src/hosted/node/scanners/scanner-theia.ts [hunk 5]
```diff
@@ -748,3 +771,7 @@ function isStringArr(something: string[]): boolean {
     return true;
 
 }
+
+function isString(something: unknown): boolean {
```
Comment (akosyakov) [inline packages/plugin-ext/src/hosted/node/scanners/scanner-theia.ts:775]: it can be inlined

File: packages/json/src/browser/json-validation-registry.ts (hunk 1 above)
Comment (bd82) [reply packages/json/src/browser/json-validation-registry.ts:31]: The reason for a separate registry is to allow applying **precedence** to the schema depending on its "origin".

See: https://github.com/eclipse-theia/theia/pull/7560/files#diff-31a8b9aa2aa16c406d15ae82f1fdc3eaR54-R61

If I understand correctly this is already happening in regards to schemas originating from the schema store as those originating from user preferences.

See: https://github.com/eclipse-theia/theia/blob/master/packages/json/src/browser/json-client-contribution.ts#L52-L56

Basically the problem is that given two schemas configs with **identical** `fileMatch` property, only the **last** one will be applied and we need a manner to define which take precedence.

Also see this comment on how it should work in VSCode:
- https://github.com/microsoft/vscode/issues/26289#issuecomment-578768254

**Comment by akosyakov** [comment]
I'm not sure about this PR. The issue is that we should drop `@theia/json` and switch to json-language-features vscode extension, after that `contributes.jsonValidation` should be implemented to work with the built-in extension.

But we can accept is as a temporary solution? cc @eclipse-theia/ecd-theia-committers

@bd82 please remove after closing of https://github.com/eclipse-theia/theia/issues/3983. It does not resolve it without switching to the built-in.

**Comment by akosyakov** [comment]
Switching to the built-in extension will automatically resolves the issue: https://github.com/theia-ide/vscode/blob/d24b5f70c69b3e75cd10c6b5247a071265ccdd38/extensions/json-language-features/client/src/jsonMain.ts#L327-L360

**Comment by bd82** [comment]
@akosyakov are there any estimated timelines for switching out `@theia/json`?

**Comment by akosyakov** [comment]
@bd82 I hope within next 2 months. We are planning to work on removing `monaco-languageclient` and getting rid of `@theia/json` is a prerequisite. cc @marcdumais-work @svenefftinge 

I think we can accept this PR for time being, but there are still 2 issues:
- whether it is fine to pull `@theia/json` via `@theia/plugin-ext`. I remember there were a PR which removed a dependency to it. @thegecko sorry if a wrong call but I remember you were concerned
- contribution shapes are bogus: https://github.com/eclipse-theia/theia/pull/7560#discussion_r408059090

**Comment by bd82** [comment]
I will fix the "wrong shapes" in the next few days

**Comment by bd82** [comment]
Hi @akosyakov I've fixed the wrong shape and another relevant CR comment.
I will now do more manual testing.

**Comment by bd82** [comment]
O.k. I've tested manually for both scenarios of fileMatch as a single string and as an array of strings. works well.

File: CHANGELOG.md [hunk 6]
```diff
@@ -14,6 +16,7 @@ Use `PluginManager.configStorage` property instead. [#7265](https://github.com/e
 
 ## v1.0.0
 
+- [core] `CommandRegistry#getAllHandlers` returns with the reversed order of handlers, so if a command has multiple handlers, any handler is considered to be more specific than the other subsequent handlers in the array. In other words, if `@theia/core` contributes a handler for a particular command and your extension also contributes a handler for the same command, the handler from your extension will have `0` index, and the handler from `@theia/core` will have `1` index when calling `getAllHandlers`.
```
Comment (akosyakov) [inline CHANGELOG.md:19]: It should not be here. How did you get it?

File: packages/plugin-ext/package.json (hunk 3 above)
Comment (akosyakov) [reply packages/plugin-ext/package.json:14]: @benoitf @tolusha It goes against https://github.com/eclipse-theia/theia/pull/6647, are you fine with it?

File: packages/json/src/browser/json-validation-registry.ts (hunk 2 above)
Comment (akosyakov) [reply packages/json/src/browser/json-validation-registry.ts:33]: @bd82 It was not addressed you this remove entry from `this.registry` and json language server should be notified about it as well

File: packages/json/src/browser/json-client-contribution.ts [hunk 7]
```diff
@@ -42,18 +43,22 @@ export class JsonClientContribution extends BaseLanguageClientContribution {
...
     }
 
     protected updateSchemas(client: ILanguageClient): void {
-        const allConfigs = [...this.jsonSchemaStore.getJsonSchemaConfigurations()];
-        const config = this.preferences['json.schemas'];
-        if (config instanceof Array) {
-            allConfigs.push(...config);
-        }
+        const schemaStoreConfigs = this.jsonSchemaStore.getJsonSchemaConfigurations();
+        const pluginContributionConfigs = this.jsonValidationContributionsRegistry.getJsonValidations();
```
Comment (akosyakov) [inline packages/json/src/browser/json-client-contribution.ts:55]: such configs can be registered dynamically, i.e. a user install or uninstall new extension. registry should fire an event and contribution should trigger `updateSchemas` then

**Comment by akosyakov** [comment]
@bd82 please see comments

File: CHANGELOG.md (hunk 6 above)
Comment (bd82) [reply CHANGELOG.md:19]: oy probably via a rebase to update my local branch

**Comment by thegecko** [comment]
> whether it is fine to pull @theia/json via @theia/plugin-ext.

I think that's OK. I seem to remember the debug package requires json anyway to correctly show the UI for configuring debug configs.

**Comment by bd82** [comment]
I will try to find time to get back to this PR next week.

File: packages/plugin-ext/package.json (hunk 3 above)
Comment (akosyakov) [reply packages/plugin-ext/package.json:14]: @azatsarynnyy Do you know whether introducing such dependency will cause issues in Che? i.e. do you use json vscode built-in extensions already?

**Comment by bd82** [comment]
Hello @akosyakov.

I am having difficulty finding the time to complete this...
I wonder if this is still relevant considering the upcoming switch to `json-language-features vscode extension`?

Is there any update /time estimates on said switch?

Thanks.
Shahar.

**Comment by akosyakov** [comment]
We've already started looking into it. The latest is mid of next month.

**Comment by bd82** [comment]
> We've already started looking into it. The latest is mid of next month.

Thanks for the update @akosyakov

**Comment by bd82** [comment]
I think this PR is now redundant so I'm closing it.
